core/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/core/KpiCalculator.kt

```diff
@@ -10,97 +10,47 @@
-    }
-
-    private fun depthFirstTraversal(
-        node: KpiHierarchyNode,
-        seen: MutableSet<KpiHierarchyNode> = mutableSetOf(),
-        action: (node: KpiHierarchyNode) -> Unit,
-    ) {
-        if (seen.contains(node)) {
-            return
+        logger.info {
+            "Running KPI calculation on $hierarchy and $rawValueKpis with strict mode=$strict"
         }
+        val root = KpiHierarchyNode.from(hierarchy.rootNode, rawValueKpis)
 
-        node.hierarchyEdges.forEach { child ->
-            depthFirstTraversal(node = child.to, seen = seen, action)
-        }
+        depthFirstTraversal(root) { it.result = calculateKpi(it, strict) }
 
-        action(node)
-        seen.add(node)
+        return KpiResultHierarchy.create(KpiHierarchyNode.to(root))
     }
 
-    private fun connectKpiHierarchyToRawValues(
-        node: KpiNode,
-        rawValueKpis: List<RawValueKpi>,
-    ): KpiHierarchyNode {
-
-        val kindToValues = mutableMapOf<KpiId, MutableList<RawValueKpi>>()
-        KpiId.entries.forEach { kindToValues[it] = mutableListOf() }
-
-        rawValueKpis.forEach { kindToValues[it.kind]!!.add(it) }
-
-        val calculationRoot = KpiHierarchyNode.from(node)
-        val updates: MutableList<TreeUpdate> = mutableListOf()
-
-        depthFirstTraversal(node = calculationRoot) { currentNode ->
-            val correspondingRawValue = kindToValues[currentNode.kpiId]
-            val parent = currentNode.parent
-
-            if (!correspondingRawValue.isNullOrEmpty() && parent != null) {
-                updates.add(TreeUpdate(parent, currentNode, correspondingRawValue))
-            }
-        }
-
-        updates.forEach {
-            val (parent, currentNode, correspondingRawValue) = it
-
-            parent.getWeight(currentNode)?.let { plannedWeight ->
-                parent.removeChild(currentNode)
-
-                val rawValueNodes =
-                    correspondingRawValue.map { rawValue ->
-                        val newNode =
-                            KpiHierarchyNode(
-                                kpiId = rawValue.kind,
-                                kpiStrategyId = KpiStrategyId.RAW_VALUE_STRATEGY,
-                                parent = parent,
-                            )
-                        newNode.setResult(rawValue.score)
-                        newNode
-                    }
-
-                if (rawValueNodes.isNotEmpty()) {
-                    val weights = plannedWeight / rawValueNodes.size
-                    rawValueNodes.forEach { rawValueNode -> parent.addChild(rawValueNode, weights) }
-                }
-            }
+    /** Selects and executes the Kpi strategy related to the given node */
+    internal fun calculateKpi(
+        node: KpiHierarchyNode,
+        strict: Boolean = false,
+    ): KpiCalculationResult {
+        logger.info { "Running KPI calculation on $node" }
```

Comment:
If for the public method log level info is OK, for the internal method (which is invoked multiple times) it should not. Otherwise we might spam logs, don't we?